lib/content.ml

```diff
@@ -1,239 +1,103 @@
-           else t, (n, `Unsafe (field, v), location) :: rest in
-         t, rest
-       | n, `Field (field, v), location ->
-         let t, rest =
-           if prefixed_by_content field
-           then { t with v= snoc (Location.inj ~location (Field { field; value= v; n; })) t.v }, rest
-           else t, (n, `Field (field, v), location) :: rest in
-         t, rest
-       | n, field, location -> t, (n, field, location) :: rest)
-    (t, []) fields
-  |> fun (t, fields) -> (t, List.rev fields)
-
-
-let fold_as_part : ((Number.t * ([> part ] as 'a) * Location.t) list) -> t -> (t * (Number.t * 'a * Location.t) list) = fun fields t ->
-  let add ~location k n v t =
-    { normalized= add k n v t.normalized
-    ; v= snoc (Location.inj ~location (Normalized { k; n; })) t.v } in
-  List.fold_left
-    (fun (t, rest) -> function
-       | n, `ContentType v, location ->
-         add ~location Key.content_type n v t, rest
-       | n, `ContentEncoding v, location ->
-         add ~location Key.content_encoding n v t, rest
-       | n, `ContentID v, location ->
-         add ~location Key.content_id n v t, rest
-       | n, `ContentDescription v, location ->
-         add ~location Key.content_description n v t, rest
-       | n, `Content (field, v), location ->
-         { t with v= snoc (Location.inj ~location (Field { field; value= v; n; })) t.v }, rest
-       | n, `Unsafe (field, v), location ->
-         let t, rest =
-           if prefixed_by_content field
-           then { t with v= snoc (Location.inj ~location (Unsafe { field; value= v; n; })) t.v }, rest
-           else t, (n, `Unsafe (field, v), location) :: rest in
-         t, rest
-       | n, `Field (field, v), location ->
-         let t, rest =
-           if prefixed_by_content field
-           then { t with v= snoc (Location.inj ~location (Field { field; value= v; n; })) t.v }, rest
-           else t, (n, `Field (field, v), location) :: rest in
-         t, rest
-       | n, field, location -> t, (n, field, location) :: rest)
-    (t, []) fields
-  |> fun (t, fields) -> (t, List.rev fields)
+  let exception Found in
+  let res : Content_encoding.t option ref = ref None in
+  try
+    Ordered.iter (fun _ (Content_field.Field (field_name, v), _) ->
+        match field_name with
+        | Content_field.Encoding -> res := Some v ; raise Found
+        | _ -> ()) t ; Content_encoding.default
+  with Found -> match !res with
+    | Some v -> v
+    | None -> assert false
+
+let parameters : t -> Content_type.Parameters.t = fun t ->
+  let exception Found in
+  let res : Content_type.Parameters.t option ref = ref None in
+  try
+    Ordered.iter (fun _ (Content_field.Field (field_name, v), _) ->
+        match field_name with
+        | Content_field.Type -> res := Some (Content_type.Parameters.of_list v.parameters); raise Found
+        | _ -> ()) t ; Content_type.Parameters.default
+  with Found -> match !res with
+    | Some v -> v
+    | None -> assert false
```

Comment:
This is really convoluted to read with the local exception -- I had to re-read it twice to satisfy myself that the assert wouldnt fire due to the default return value in Ordered.iter.  Is it possible to factorise out the local exception into a helper function perhaps?

Reply:
Ok, may be a more functional way (like a `fold`) should be better.